File: 剑指Offer/5.替换空格.cpp
class Solution {
public:
    string replaceSpace(string s) {
        // 先统计空格数，然后倒着替换
        int numberOfBlanks = 0;
        for (char c : s) {
            if (c == ' ') numberOfBlanks++;
        }
        int totalLength = s.length() + numberOfBlanks * 2;
        char newStr[totalLength + 1];
        newStr[totalLength] = '\0';

        int index = totalLength - 1;
        for (int i = s.length() - 1; i >= 0; --i) {
            if (s[i] != ' ') {
                newStr[index--] = s[i];
            }
            else {
                newStr[index--] = '0';
                newStr[index--] = '2';
                newStr[index--] = '%';
            }
        }
        return newStr;
    }
};

File: 剑指Offer/5.替换空格.py
class Solution:
    def replaceSpace(self, s: str) -> str:
        numberOfBlanks = 0
        for c in s:
            if c == ' ':
                numberOfBlanks += 1

        totalLength = len(s) + numberOfBlanks * 2
        newStr = [' '] * totalLength
        # newStr = ' ' * totalLength
        index = totalLength - 1

        for i in range(len(s) - 1, -1, -1):
            if s[i] != ' ':
                newStr[index] = s[i]
                index -= 1
            else:
                newStr[index] = '0'
                newStr[index - 1] = '2'
                newStr[index - 2] = '%'
                index = index - 3

        return ''.join(newStr)